inc/class-expiring-posts.php

```diff
@@ -0,0 +1,285 @@
+<?php
+/**
+ * Expiring_Posts class file
+ *
+ * @package Expiring_Posts
+ */
+
+namespace Expiring_Posts;
+
+use InvalidArgumentException;
+use WP_Post;
+
+/**
+ * Expiring Posts Manager
+ */
+class Expiring_Posts {
+	/**
+	 * Cron hook to run expiration check.
+	 *
+	 * @var string
+	 */
+	const CRON_HOOK = 'expiring_posts_check_expiration';
+
+	/**
+	 * Instance of the manager.
+	 *
+	 * @var static
+	 */
+	protected static $instance;
+
+	/**
+	 * Post types that are registered for expiration.
+	 *
+	 * @var array<string, array>
+	 */
+	protected array $post_types = [];
+
+	/**
+	 * Get the instance of this singleton
+	 *
+	 * @return static
+	 */
+	public static function instance() {
+		if ( ! isset( static::$instance ) ) {
+			static::$instance = new static();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Constructor.
+	 */
+	protected function __construct() {
+		add_action( static::CRON_HOOK, [ $this, 'run_expiration_check' ] );
+
+		$this->schedule_next_run();
+	}
+
+	/**
+	 * Register a post type that should automatically expire.
+	 *
+	 * @throws InvalidArgumentException Thrown on invalid action.
+	 * @throws InvalidArgumentException Thrown on invalid expire_after.
+	 *
+	 * @param string $post_type Post type to register.
+	 * @param array  $args {
+	 *     Arguments to add the post type.
+	 *
+	 *     @type string $action       Action to apply to the post upon expiration: draft/trash/delete.
+	 *     @type int    $expire_after Number of seconds for the post to be expired after,
+	 *                                defaults to a year.
+	 * }
+	 */
+	public function add_post_type( string $post_type, array $args ) {
+		if ( ! post_type_exists( $post_type ) ) {
+			_doing_it_wrong(
+				__FUNCTION__,
+				sprintf(
+					/* translators: 1: post type */
+					esc_html__( 'Post type does not exist for expiration: %s', 'expiring-posts' ),
+					esc_html( $post_type ),
+				),
+				'0.1.0',
+			);
+
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			[
+				// Action to apply to the post after it is expired (draft/trash).
+				'action'       => 'draft',
+				// Number of seconds for the post to be expired after.
+				'expire_after' => YEAR_IN_SECONDS,
+			],
+		);
+
+		// Validate the arguments for the post type.
+		if ( ! in_array( $args['action'], [ 'delete', 'draft', 'trash' ], true ) ) {
+			throw new InvalidArgumentException(
+				sprintf(
+					/* translators: 1: action */
+					__( 'Invalid action for expiration (expected delete/draft/trash): %s', 'expiring-posts' ),
+					$args['action'],
+				),
+			);
+		}
+
+		// Validate the expire after value.
+		if ( ! is_int( $args['expire_after'] ) ) {
```

Comment:
I wonder if we should check if it’s greater than 0 too? A negative integer could screw things up.

Reply:
Good point !